docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,727 @@
+documents, but these are not too onerous.
+
+### Change Commits
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+```
+.
+|--change
+|--manifest.toml
+```
+
+This tree contains a single change to a collaborative object. We will go into
+more details shortly. Any direct dependencies of this change are encoded in the 
+same manner and become the parents of this commit. This allows us to
+reconstruct the automerge depdency graph. 
+
+Along with the dependencies of the commit we also need to add the commit of the
+identity which created this commit. We need this identity to validate
+signatures and by making the commit a parent we ensure that git will replicate
+it for us. 
+
+A valid change commit must have three trailers:
+
+- `X-Rad-Signature`, as for identity documents
+- `X-Rad-Author-Parent`, this is the hash of the commit which references the
+  author identity. We use this trailer to avoid following the author commit
+  reference when constructing the automerge change graph
+- `X-Rad-Schema-Parent`, this is the hash of the parent commit which contains 
+  the schema of this object. See [schema commits](#schema-commits).
+
+
+#### `manifest.toml`
+
+The manifest is a TOML file containing some metadata about the object.
+Specifically it will contain:
+
+- `id`, a UUID, generated at the time the object is created
+- `typename`, discussed above
+- `history_type`, always `"automerge"`, this is here to allow for different
+  CRDT implementations in future.
+
+
+Each object is also created with a JSON schema. The schema is represented by an
+initial `schema.json` and a series of schema migrations which extend that
+initial schema. Schema migrations will not be addressed in detail in this RFC
+but we will show their feasibility.
+
+#### `change`
+
+This is the automerge change which this commit introduces. It is a binary file
+which must contain a single change and it's dependents must be the dependents
+referenced by the parents of the commit.
+
+
+### Reconstructing Collaborative Objects
+
+Assuming we have replicated a number of collaborative objects from our tracking
+graph, we can now view the merged state of those objects. To do this we search
+through every `/rad/collaborative-objects/<typename>/<object ID>` reference for
+every remote we have and collect the change files for each object ID.
+
+At this point we have the hash linked graph of automerge changes, but we need
+to make sure that the merged document is authenticated and valid with respect
+to it's schema. To do this we start at the roots of the hash graph and walk
+down the tree. As we encounter each change we check it's signature, apply it
+and check that the new document does not violate the schema. If it does violate
+the schema we discard the change and all dependent changes. Finally, we have an
```

Comment:
This might be tricky with concurrent edits.

For example we could have a document `{ foo: ["x"] }` where the length of the array at `foo` must be less than 3. If two peers concurrently add an element to produce `{ foo: ["x", "y"] }` and `{ foo: ["x", "z"] }` then each individual change is ok. But as soon as a peer sees both of them and combines them they become invalid. Do we reject both of the changes now? This might allow people to sabotage documents.

I’d also like to know more about whether we want to reject updates at the gossip layer. I.e. do we forward messages that produce invalid documents?

Reply:
Perhaps JSON schema constraints are only applicable insofar as the underlying CRDT can guarantee them? As in, perhaps we need to define a vocabulary for the kinds of constraints which are valid?

> I’d also like to know more about whether we want to reject updates at the gossip layer. I.e. do we forward messages that produce invalid documents?

We do not have individual messages, so the only choice would be to stop replicating the entire history of the violating peer from the invalid commit on. Given that the violation may be because of a bug or unsound constraint, I'd rather leave this decision to the human. I also don't think that there should be any unattended eager evaluation: there could be errors, conflicts, or even timetravel due to newly-discovered remotes. All of these need to be surfaced to the user somehow, so evaluation would more naturally occur on read.

Reply:
I'm currently working on putting together an alternative JSON schema vocabulary because I think that's the most principled way to approach this. We would choose a subset of [JSON schema validation](http://json-schema.org/draft/2019-09/json-schema-validation.html) (the default vocabulary) which only includes constraints that are commutative (e.g `oneOf` for values, but not `maxItems` for arrays). This vocabulary would also include the schema extensions we will need for authenticating data structures (e.g `rad_signed_by` etc.).